File: py/381_insert_delete_getrandomO(1)_dup.py
import random
class RandomizedCollection(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.map = {}
        self.vec = []

    def insert(self, val):
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        res = val in self.map
        idx = len(self.vec)
        if res:
            self.map[val].append(idx)
            self.vec.append(val)
        else:
            self.map[val] = [idx]
            self.vec.append(val)
        return not res

    def remove(self, val):
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        :type val: int
        :rtype: bool
        """
        res = val in self.map
        if res:
            idx = self.map[val][-1]
            if idx != len(self.vec) - 1:
                num_back = self.vec[-1]
                self.map[num_back].remove(len(self.vec) - 1)
                self.vec[-1], self.vec[idx] = self.vec[idx], self.vec[-1]
                self.map[val].pop()
                if len(self.map[val]) == 0:
                    del self.map[val]
                self.vec.pop()
                self.map[num_back].append(idx)
            else:
                self.map[val].pop()
                if len(self.map[val]) == 0:
                    del self.map[val]
                self.vec.pop()
        return res

    def getRandom(self):
        """
        Get a random element from the collection.
        :rtype: int
        """
        return random.choice(self.vec)

if __name__ == '__main__':
    collection = RandomizedCollection()
    print collection.insert(0)
    print collection.insert(1)
    print collection.insert(2)
    print collection.insert(3)
    print collection.insert(3)
    print collection.remove(2)
    print collection.remove(3)
    print collection.remove(0)
    print collection.getRandom()
    print collection.getRandom()
    print collection.getRandom()

File: leetcode_cpp/381_insert_delete_getrandom_O(1)_DUP.cpp
//
// Created by Dicky Shi on 6/13/17.
//
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <ctime>
#include "utils.h"
using namespace std;

class RandomizedCollection {
public:
    /** Initialize your data structure here. */
    RandomizedCollection() {
        srand((unsigned)time(NULL));
    }

    /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
    bool insert(int val) {
        bool ret = false;
        if (num_map[val].empty()) {
            vector<int> n;
            num_map[val] = n;
            ret = true;
        }
        num_map[val].push_back(nums.size());
        nums.push_back(val);
        return ret;
    }

    /** Removes a value from the collection. Returns true if the collection contained the specified element. */
    bool remove(int val) {
        if (num_map[val].empty()) {
            return false;
        } else {
            int idx = num_map[val].back();
            if (idx != nums.size() -1) {
                int num_back = nums.back();
                num_map[num_back].erase(find(num_map[num_back].begin(), num_map[num_back].end(), nums.size()-1));
                num_map[num_back].push_back(idx);
                swap(nums[idx], nums.back());
            }
            num_map[val].pop_back();
            nums.pop_back();
            return true;
        }
    }

    /** Get a random element from the collection. */
    int getRandom() {
        int r = rand() % nums.size();
        return nums[r];
    }

private:
    unordered_map<int, vector<int>> num_map;
    vector<int> nums;
};

int main() {
//    clock_t start = clock();
    RandomizedCollection *obj = new RandomizedCollection();
    cout << obj->insert(0) << endl;
    cout << obj->insert(1) << endl;
    cout << obj->remove(0) << endl;
    cout << obj->insert(2) << endl;
    cout << obj->remove(1) << endl;
    cout << obj->getRandom() << endl;
//    clock_t end = clock();
//    cout << "Time used: " << end - start << endl;
    return 0;
}
